### PR TITLE
CLI: Improve monorepo support in `automigrate` and revamp framework-related migrations

### LINKED ISSUE #20508: [Feature Request]: Accept --configFolderPath option in automigrators & migrators
### Is your feature request related to a problem? Please describe

We want to be able to call the automigrators from Nx. In Nx, there are multiple `.storybook` directories, and multiple `.storybook/main|preview.js` files.

From within a Nx generator, we want to recursively call the automigrators (eg. `npx sb automigrate mainjsFramework`) and also pass a flag to tell Storybook which directory to look into, so for example:

```
npx sb automigrate --fixId mainjsFramework -c apps/my-app/.storybook
```

That would enable us to call the automigrators from Nx, or allow anyone to call the automigrators for a Nx workspace.

### Describe the solution you'd like

The functions that are listed [here](https://github.com/storybookjs/storybook/blob/next/code/lib/cli/src/automigrate/fixes/index.ts) to accept an option for `configFolderPath` (or anything similarly named, to allow Nx to pass the directory).

Similarly, if there's a migrator that looks for any other Storybook-related files (eg. stories transformation) to accept a relevant `glob` or something (like `sb migrate csf-2-to-3` which accepts `--glob`).

### Describe alternatives you've considered

_No response_

### Are you able to assist to bring the feature to reality?

yes, I can

### Additional context

_No response_

### LINKED ISSUE #20955: [Feature Request]: Improve how automigrations read main.js
### Is your feature request related to a problem? Please describe

The Storybook automigrations follow this pattern:
1. **check** (do assertions to know whether the automigration is needed)
2. **prompt** (ask user if they want migration to be done)
3. **run** (perform a migration on user's behalf)

Currently, we use AST based tooling (csf-tools) to both **read** main.js values in the _check_ step, and also to **manipulate** main.js in the _run_ step.

### The problem

The way we read main.js by using AST analysis is error-prone given that users might have an unconventional format of their main.js files, e.g.

```js
import baseConfig from '../storybook-utils/base-config'
export default baseConfig
```

```js
import baseConfig from '../storybook-utils/base-config'
export default {
   ...baseConfig,
   stories: []
}
```

```js
import baseConfig, { mergeConfig } from '../storybook-utils/base-config'
export default mergeConfig(baseConfig, {
   ...baseConfig,
   stories: []
})
```

Here's an example of an issue caused by it: #20940 (more users got affected by it)

### Describe the solution you'd like

The automigrations should be improved in two ways:

### Evaluating main.js with `presets.apply` in the _run_ step

This way, it doesn't really matter what the main.js looks like. We will always be able to evaluate what we need. This comes with a small performance cost, and possibility of having some issues e.g. crashing if automigrations run in a project that has not installed dependencies.

### Adding warnings in the _check_ step

If users have a non AST friendly main.js, we should make sure that the automigrations don't crash, but rather tell the users that it was not possible to be done **because their main.js is unconventional and we recommend changing**, and provide info for them to do the migration manually themselves

### PR DESCRIPTION
**Pull request by mandarini**
closes #20508, closes #20955

### new CLI flags added to `storybook automigrate`

Automigrations now accept `--config-dir` that points to the `.storybook` directory that is to be transformed, and `--renderer` that has the renderer package of the project in use (e.g. @storybook/react, @storybook/angular). Both are optional. They are useful for monorepo-like scenarios, such as Nx, which will use these flags to migrate the existing projects that have storybook configuration.

Additionally, the `--skip-install` flag was also added.

For users in a monorepo that run the command without passing --config-dir, the migrations will stop with the following message:
image

The `renderer` flag is used to specify which framework the project is using. If this is not passed, and the workspace has multiple renderers installed, the user will get prompted:
image

### Revamped automigrations

The following automigrations are now merged into a single one:
- mainjsFramework
- new-frameworks
- nextjs-framework
- sveltekit-framework

This way, users will get from zero to the correct point without having to run multiple migrations that tell them things like "you have a wrong setup, please upgrade to X. Well, actually.. you have a wrong setup, please upgrade to Y".

It's incremental, and there are dozens of possibilities to trigger the migration, where each step is shown like so:

image

The automigration also **does not** use AST parsing for _reading_ main.js, which makes them support non-standard main.js files like so:
```js
import getConfig from './baseConfig'
export default getConfig()
```

The automigration **will use** AST parsing for _writing_ into main.js when the user prompts that they want the migration to be done, and if there are failures, they will get a helpful message like so:
image

### Changes to AST tool: ConfigFile

Currently, the csf-tools helpers write a `named export` into a main.js file, even if it has `default export`. This is error prone and makes the end result incorrect. This PR makes that scenario throw an error instead.

### Misc

This PR adds a `jest_workaround` library to configure jest/swc given that the swc implementation has a limitation for mocking, which is solved by that dependency.

## TODO

- [x] Update MIGRATIONS.md

## How to test

There are lots of unit tests, please check them.

1. Go into a "migratable" Storybook project
2. Try to migrate it with `sb automigrate`
3. Try to migrate it with `sb automigrate` in a monorepo-like scenario (multiple renderers in package.json) and get prompted
4. Try to migrate it with `sb automigrate` in a monorepo-like scenario using the --renderer flag and do not get prompted
5. Try to migrate it with `sb automigrate` using --config-dir flag in a custom storybook config-dir scenario

## Checklist

- [x] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [x] Make sure to add/update documentation regarding your changes
- [x] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [x] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [x] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
@mandarini did you test this?

It looks great to me.

**Review by shilman — COMMENTED** [review]
@mandarini Thank you for this PR!

A couple questions:
1. What do you think about renaming `frameworkPackage` to `framework` for consistency?
2. The example you give in your PR description uses `@storybook/react` as a framework. However in 7.0, `@storybook/react` is a "renderer" and `@storybook/react-webpack5/vite/etc.` is a "framework". Do you actually mean renderer here?

**Comment by mandarini** [comment]
@shilman the reason I named it `frameworkPackage` was actually for consistency. [I saw in the code](https://github.com/storybookjs/storybook/blob/next/code/lib/core-common/src/utils/get-storybook-info.ts#L54) that `framework` is the name of the framework without the `@storybook/`. If on Storybook 7 the `@storybook/react` is a renderer package, then I can rename it to [`rendererPackage`](https://github.com/storybookjs/storybook/blob/next/code/lib/core-common/src/utils/get-storybook-info.ts#L39). What do you think?

**Comment by shilman** [comment]
@yannbf I think you looked at this part of the code recently and wanted to change how we dealt with frameworks/renderers. Any thoughts on how we might update that & whether we should do it before or after merging this PR?

**Comment by yannbf** [comment]
Let me take a look at this this week! I'll add the "do not merge" tag so we don't merge this accidentally!
I'll reach out @mandarini

**Comment by ndelangen** [comment]
Thanks @yannbf you're the best

**Comment by yannbf** [comment]
Hey @mandarini I took a look at this, but definitely need to take a deeper look into the whole use case for this PR.
I think we need to process the use case a little more, and provide a robust solution for multi `main.js` files, otherwise as we make future automigrations, we might potentially have trouble in monorepo scenarios without noticing.

I wonder if we could make a solution where the automigration command knows it's running in a monorepo setting (either automatically or via a flag), and from its current working directory, it finds a main.js, but it also "finds up" another main.js (I assume there are no scenarios where a user might have inheritance of 3 main.js files), and we can do assertions based on a combination of those main.js nodes. This way there's no need to specify a configDir or a framework as flags.

I don't know, what I do know is that we need to come up with a robust solution to handle multi main.js configs, and have it available for any of our static analysis tools that need it (CLI, eslint-plugin-storybook, etc)

Thoughts? I think @valentinpalkovic and @shilman will be interested in discussing about this too

**Comment by mandarini** [comment]
@yannbf we can discuss this more, sure. Finding all `main.js` files sounds great. The way i would do it in Nx, for example, I would look for all the projects that use Storybook, and each of these projects specifies the storybook `configDir` (which is already available in the Storybook CLI options - the "Directory where to load Storybook configurations from"). Then I would run the `npx sb automigrate` command using that configDir.

We can find another way to do it, sure. My intention was to leave the "logic of finding all the Storybook configDirs" up to Nx, but if automigrate ends up doing that, even better. I thought that just passing a "configDir" flag would be less intrusive, now. 

The other issue is with the framework package. Right now, automigrate looks into the `package.json` and assumes that the framework used is the first `@storybook/*` framework dependency it finds, which does not work in multi-project setups (since you can have angular, react, etc). So, we should come up with a more robust way of determining the framework used (looking at the `builder` in combination with something else)? In the case of Nx, we just have the `uiFramework` option set up in `project.json`, so I take it from there, and then I would pass it to `automigrate` using that flag.

Let me know what you think and we can have a call about it if you want.

**Comment by mandarini** [comment]
Thank you @yannbf ! I will test this out tomorrow!

File: code/.swcrc
Comment (yannbf) [inline code/.swcrc:20]: this change is needed as I'm parsing this file as JSON in `code/jest.config.base.js`

File: code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts
Comment (yannbf) [inline code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts:49]: I'm adding a comment here so this file doesn't get overlooked. Github is collapsing it by default

File: code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.test.ts
Comment (yannbf) [inline code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.test.ts:46]: I'm adding a comment here so this file doesn't get overlooked. Github is collapsing it by default

**Review by ndelangen — APPROVED** [review]

File: code/lib/cli/src/automigrate/fixes/new-frameworks/utils.ts
Comment (valentinpalkovic) [inline code/lib/cli/src/automigrate/fixes/new-frameworks/utils.ts:41]: Do we not already have somewhere a map like this? If not, can we put it somewhere where it is more accessible (not in `automigrate`)?

File: code/lib/cli/src/automigrate/fixes/new-frameworks/utils.ts
Comment (valentinpalkovic) [inline code/lib/cli/src/automigrate/fixes/new-frameworks/utils.ts:54]: I'm not too fond of this inline condition for extensibility reasons. What if we introduce a new builder? This code will very likely break.

Instead let's do it like this:

```tsx
type Builders = 'vite' | 'webpack5';

export const getBuilderInfo = (...): { name: Builders; options: any } => {

  let builderName: Builders | null = null;
  
  if (builder) {
    if (typeof builder === 'string') {
      if(builder.includes('vite')) {
        builderName = 'vite'
      }
    ...
  if(builderName === null) {
    throw Error("Builder couldn't be extracted")
  }
}

File: code/lib/cli/src/automigrate/helpers/detectFramework.ts
Comment (valentinpalkovic) [inline code/lib/cli/src/automigrate/helpers/detectFramework.ts:15]: I know it's an edge case, but what happens if `matches` is 0. Do we want to let the user know, that his framework is not supported yet or we couldn't look up a framework?

File: code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts
Comment (valentinpalkovic) [inline code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts:60]: I saw that multiple times. Can we extract this into the `packageManager` and call a method instead to get all dependencies?

File: code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts
Comment (valentinpalkovic) [inline code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts:72]:
```suggestion
        ❌ Unable to determine Storybook version.
```

File: code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts
Comment (valentinpalkovic) [inline code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts:152]:
```suggestion
    let addonOptions: Record<string, any> = {};
```

File: code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts
Comment (valentinpalkovic) [inline code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts:155]: If you want to go the extra mile, you could warn Next.js users below 12.2.0 that not the whole feature set is supported (e.g. `next/future/image`)

File: code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts
Comment (valentinpalkovic) [inline code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts:346]: Subtle, but please try to be consistent about spacing: `Webpack 4`, `Webpack4`, `Webpack 5`

File: code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts
Comment (yannbf) [reply code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts:60]: Maybe we can do that in the future, in a separate PR! If you feel strong about it, could you create an issue?

File: code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts
Comment (yannbf) [reply code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts:155]: I think that's a little too much hehe

File: code/lib/cli/src/automigrate/fixes/new-frameworks/utils.ts
Comment (yannbf) [reply code/lib/cli/src/automigrate/fixes/new-frameworks/utils.ts:43]: This is just useful for the automigration. However, this PR exposes two maps: `rendererPackages` and `frameworkPackages` that list all renderers and all frameworks (that we know of) in `@storybook/core-common`

**Review by valentinpalkovic — APPROVED** [review]

File: MIGRATION.md
Comment (shilman) [inline MIGRATION.md:452]: why did the order change? alphabetical better?

File: MIGRATION.md
Comment (shilman) [inline MIGRATION.md:559]: @yannbf so is `core.builder` completely ignored in SB7?

File: code/lib/core-common/src/utils/validate-configuration-files.ts
Comment (shilman) [inline code/lib/core-common/src/utils/validate-configuration-files.ts:25]: @yannbf Do we still need this now that there is the global pre-check? (applies to every instance of this line in this PR)

**Review by shilman — APPROVED** [review]
A few misc questions/comments, but otherwise looks great!! 🚀

File: code/lib/core-common/src/utils/validate-configuration-files.ts
Comment (yannbf) [reply code/lib/core-common/src/utils/validate-configuration-files.ts:25]: Yes, as `validate-configuration-files` can run in a non-automigrate scenario (anywhere where loadMainConfig is called)

File: MIGRATION.md
Comment (yannbf) [reply MIGRATION.md:452]: I made it close to react as nextjs is a react thing, so people looking for react would quickly notice it. Happy to change this!

File: MIGRATION.md
Comment (yannbf) [reply MIGRATION.md:559]: No it isn't. SB reads `framework.options.builder` first, and if not there, it reads `core.builder.options`. You can see the logic [here](https://github.com/storybookjs/storybook/blob/feat/configDir-automigrations/code/lib/core-common/src/utils/get-builder-options.ts#L7)

I'd argue it's better to just tell people to move their builder stuff to framework, and have all users follow that pattern, with the only exception of a non-standard user config, which would use a third party builder or something that won't match the framework somehow. In that case though, the user probably would not follow this particular instruction as the first step is to pick a framework from the list of known frameworks we provide

**Comment by mandarini** [comment]
THANK YOUUUUUU!!!! :D :D :D

**Comment by mandarini** [comment]
Just tested it, too, and works like a charm!!!! :D

**Pull request merged**
